data_model/src/predicate.rs

```diff
@@ -46,10 +74,83 @@ impl PredicateBox {
             Self::And(preds) => Self::Or(preds.into_iter().map(Self::negate).collect()),
             Self::Or(preds) => Self::And(preds.into_iter().map(Self::negate).collect()),
             Self::Not(pred) => *pred, // TODO: should we recursively simplify?
-            Self::Raw(pred) => Self::Not(Box::new(PredicateBox::Raw(pred))),
+            Self::Raw(pred) => Self::Not(Box::new(Self::Raw(pred))),
+        }
+    }
+}
+
+pub trait PredicateTraitExt<I>: Sized {
+    fn and(self, left: Self) -> GenericPredicateBox<Self> {
+        GenericPredicateBox::and(self, left)
+    }
+
+    fn or(self, left: Self) -> GenericPredicateBox<Self> {
+        GenericPredicateBox::or(self, left)
+    }
+
+    fn negate(self) -> GenericPredicateBox<Self> {
+        GenericPredicateBox::negate(self.into())
+    }
+}
+
+impl<T, I> PredicateTraitExt<I> for T
+where
+    T: PredicateTrait<I>,
+    I: Copy,
+{
+}
+
+impl<T> From<T> for GenericPredicateBox<T> {
+    fn from(pred: T) -> Self {
+        GenericPredicateBox::new(pred)
+    }
+}
+
+impl<T, I> PredicateTrait<I> for GenericPredicateBox<T>
+where
+    I: ?Sized + Copy,
+    T: PredicateTrait<I>,
+{
+    type EvaluatesTo = T::EvaluatesTo;
+
+    #[inline] // This is not a simple function, but it allows you to inline the logic and optimise away the logical operations.
+    fn applies(&self, input: I) -> Self::EvaluatesTo {
+        match self {
+            Self::Raw(predicate) => predicate.applies(input),
+            Self::And(predicates) => {
+                let mut operands = predicates.iter().map(|predicate| predicate.applies(input));
+                let mut value = operands.next().unwrap();
+                for operand in operands {
+                    if operand.is_truthful() {
+                        value = value.and(operand);
```

Comment:
Not sure we need this specifically. It looks to me that we're getting a little too caught up with not knowing whether we're dealing with ternary or binary logic.